### ISSUE TITLE
`mem_alloc(): Cannot allocate memory` with huge wordlist and high number of forks

### ISSUE THREAD

**Issue by AlekseyCherepanov** [openwall/john#5126]
I got access to an unused machine with 512Gb of memory and 96 threads. I have non-openmp build. When I use `--fork=96` with rules and a huge wordlist (like HIBP cracks), I get messages like the following:
```
mem_alloc(): Cannot allocate memory trying to allocate 8000000008 bytes
```

Wordlist without rules works well.

Preparations:
```
$ echo '$MD4$aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' > md4.pw
$ seq 1000000000 > s1g.txt
$ ls -sh s1g.txt
9.3G s1g.txt
$ free -h
              total        used        free      shared  buff/cache   available
Mem:          503Gi       1.1Gi       502Gi       3.0Mi       138Mi       500Gi
Swap:         974Mi        63Mi       911Mi
```

Example of failure. There were 8 `mem_alloc()` lines, 29 `DONE` lines:
```
$ ./john/run/john md4.pw --fork=96 --wordlist=s1g.txt --rules=':'
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD4 [MD4 512/512 AVX512BW 16x3])
Node numbers 1-96 of 96 (fork)
Each node loaded the whole wordfile to memory
mem_alloc(): Cannot allocate memory trying to allocate 8000000008 bytes
[...]
mem_alloc(): Cannot allocate memory trying to allocate 8000000008 bytes
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
Enabling duplicate candidate password suppressor
67: posix_memalign (268435456 bytes): Cannot allocate memory
94 0g 0:00:03:57 DONE (2022-05-02 17:51) 0g/s 43929p/s 43929c/s 43929C/s 999899230..999999934
[...]
Waiting for 95 children to terminate
41 0g 0:00:04:09 DONE (2022-05-02 17:52) 0g/s 41701p/s 41701c/s 41701C/s 999899177..999999977
Main process session completed, but some child processes failed
```

`--dupe-suppression=0` does not help.

`--fork=65` works well. `--fork=66` fails. 66 prints this message unlike 65:
```
Each node loaded the whole wordfile to memory
```

I know that `mem_alloc()` failure exits respective fork. But message `Main process session completed, but some child processes failed` is printed only in the end. So for longer attacks, message about `mem_alloc()` is not very informative. Maybe a word "error" / "failure" / "critical" would help.

`--fork=960` with 848Mb wordlist does not fail. It does not print `Each node loaded the whole wordfile to memory`.

So I guess logic to read the whole wordlist into memory might be wrong for some cases.

**Comment by solardiz** [comment]
I thought this would be some kind of integer overflow, but I don't see that in our latest code. Puzzling.

**Comment by magnumripper** [comment]
Just a quick comment without yet trying to understand the whole issue: The code path is the exact same for MPI as for `--fork`. If running under MPI, we have (up to) 96 different machines while a `--fork` run is all on one machine. We don't really know which it is, so some decisions may end up suboptimal for the actual situation.

**Comment by AlekseyCherepanov** [comment]
I played with `--mem-file-size=`. My guess about the logic: size of dictionary is divided by number of forks, after that if result is below value for `--mem-file-size=` then the whole file is loaded into memory. The logic seems wrong.

Default is 150 MB:
```
$ ./john/run/john --help
[...]
--mem-file-size=SIZE       Size threshold for wordlist preload (default 150 MB)
```

Respective expression in `options.c` and constant in `params.h`:
```c
WORDLIST_BUFFER_DEFAULT / 1000000

#define WORDLIST_BUFFER_DEFAULT		150000000
```

It describes the difference between `--fork=65` (above the limit) and `--fork=66` (below the limit). Exact size of `s1g.txt` is 9888888899.
```
>>> 9888888899 / 65
152136752
>>> 9888888899 / 66
149831649
```

Dividing by 96: 103009259.
```
$ ./john/run/john md4.pw --fork=96 --wordlist=s1g.txt --rules=':' --mem-file-size=103009259
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD4 [MD4 512/512 AVX512BW 16x3])
Node numbers 1-96 of 96 (fork)
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
Enabling duplicate candidate password suppressor
[... works well ...]

$ ./john/run/john md4.pw --fork=96 --wordlist=s1g.txt --rules=':' --mem-file-size=103009260
Using default input encoding: UTF-8
Loaded 1 password hash (Raw-MD4 [MD4 512/512 AVX512BW 16x3])
Node numbers 1-96 of 96 (fork)
Each node loaded the whole wordfile to memory
[... failure ...]
```

This time the main process failed. So I got back to shell's prompt and then more failures showed up. Later DONE status lines appeared. There was not the message `Main process session completed, but some child processes failed`. Also there was not the message about the suppressor.

**Comment by AlekseyCherepanov** [comment]
So a lower value for `--mem-file-size=` is a workaround for the problem.

**Comment by solardiz** [comment]
Aleksey, per what you wrote it looks like the size of a would-be portion of a wordlist is compared against the threshold even when the whole wordlist would then be loaded. If so, that's the logic error - we should be consistent in what size we compare and what size we load.

**Comment by AlekseyCherepanov** [comment]
I agree that the logic error should be fixed. I meant a lower value for `--mem-file-size=` (e.g. `--mem-file-size=1`) might be a suitable workaround for user with such problem before the real fix.

**Comment by solardiz** [comment]
Maybe fix (untested):

```diff
+++ b/src/wordlist.c
@@ -632,10 +632,14 @@ void do_wordlist_crack(struct db_main *db, const char *name, int rules)
                }
 #endif
 
-               ourshare = options.node_count ?
-                       (file_len / options.node_count) *
-                       (options.node_max - options.node_min + 1)
-                       : file_len;
+               ourshare = file_len;
+
+               // Load only this node's share of words to memory
+               if (mem_map && options.node_count > 1 &&
+                   (file_len > options.node_count * (length * 100))) {
+                       ourshare = (file_len / options.node_count) *
+                               (options.node_max - options.node_min + 1);
+               }
 
                if (ourshare < options.max_wordfile_memory &&
                    mem_saving_level < 2 &&
@@ -649,8 +653,7 @@ void do_wordlist_crack(struct db_main *db, const char *name, int rules)
                        char *aep;
 
                        // Load only this node's share of words to memory
-                       if (mem_map && options.node_count > 1 &&
-                           (file_len > options.node_count * (length * 100))) {
+                       if (ourshare < file_len) {
                                /* Check net size for our share. */
                                for (nWordFileLines = 0;; ++nWordFileLines) {
                                        char *lp;
```

I don't understand why we have this depend on `mem_map` like that, but that's a separate issue.

**Comment by solardiz** [comment]
@AlekseyCherepanov Can you give my patch above a try, repeating the same tests you ran previously? Thank you!

**Comment by AlekseyCherepanov** [comment]
It seems ok:
```
$ test/run/john md4.pw --fork=96 --wordlist=s1g.txt --rules=':'
[... ok ]

$ test/run/john md4.pw --fork=960 --wordlist=s100m.txt --rules=':'
[... ok as it was ]


$ ls -l s100m.txt
[...] 888888898 [...]
$ test/run/john md4.pw --fork=96 --wordlist=s100m.txt --mem-file-size=1000000000 --rules=':'
[...]
Each node loaded 1/96 of wordfile to memory (about 8 MB/node)
[... ^ the fix does not affect it ]

$ test/run/john md4.pw --fork=96 --wordlist=s1g.txt --mem-file-size=10000000000 --rules=':'
Each node loaded the whole wordfile to memory
[... failures as expected because 10GB is specified for --mem-file-size= ]


$ # exact size
$ test/run/john md4.pw --fork=2 --wordlist=s1g.txt --mem-file-size=9888888899 --rules=':'
[... ok without "loaded" ]

$ # size + 1
$ test/run/john md4.pw --fork=2 --wordlist=s1g.txt --mem-file-size=9888888900 --rules=':'
[...]
Each node loaded the whole wordfile to memory
[... ok because fork=2 ]
```

**Comment by solardiz** [comment]
Thanks, Aleksey. I think I'll also fix the off-by-one and create a PR.

**Comment by magnumripper** [comment]
> I don't understand why we have this depend on `mem_map` like that, but that's a separate issue.

Long ago, we actually loaded the wordlist (or our share of it) in a contiguous block of memory, then (if rules in use) we made an array to immediately find any word by index.

Nowadays, we normally memory map the wordlist instead (so any number of forks will share the one same copy in memory) but we still do that indexing thing - and it's assumably the latter that runs out of memory in this issue.

**Comment by AlekseyCherepanov** [comment]
> Nowadays, we normally memory map the wordlist instead (so any number of forks will share the one same copy in memory) but we still do that indexing thing - and it's assumably the latter that runs out of memory in this issue.

`8000000008` in the message is `8 * (1G + 1)` while the wordlist has 1G lines. But `s1g.txt` is not mapped, while `s100m.txt` is mapped (I don't have `strace` there, so I checked `/proc/<pid>/maps`).

Index can be shared across forks too, cannot it?

**Comment by solardiz** [comment]
> that indexing thing - and it's assumably the latter that runs out of memory in this issue.

I think in Aleksey's case with 8000000008 the system ran out of memory because both the wordlist was read into memory (read, not mapped) by each node, and the index was created by each node.

> Index can be shared across forks too, cannot it?

Yes, but we reach the code that creates it only after we've already forked. So this is not easy to implement now, and I doubt we even need that index - instead (after we release 2.0) we should rework `wordlist.c` and just seek to offset on restore, like core always did.

**Comment by AlekseyCherepanov** [comment]
Oh, indeed it looks like the not mapped wordlist is not shared across forks. I tried 300M, 310M and 320M candidates with `--fork=96` and high `--mem-file-size=` to load the whole dictionary: 300M was ok, 310M failed one time, 320M failed reliably.

Size of 320M is 3088888898. The wordlist loaded by each fork plus index is slightly over the capacity. While index alone is not close to the capacity.
```
>>> 310000000 * 8 * 96
238080000000
>>> 3088888898 * 96 + 320000000 * 8 * 96
542293334208
527832252  <-- total memory reported by `free`
```

**Comment by magnumripper** [comment]
> > Index can be shared across forks too, cannot it?
> 
> Yes, but we reach the code that creates it only after we've already forked. So this is not easy to implement now

For the case where each node indexes only its own words, there's nothing to share. Not sure why this did not happen here? The code is a mess and it handles too many different situations: Rules or not, memory map or not, index or not, loopback or not. Node splitting the words **or** the rules.

Also, the MPI case use [nearly] the exact same code path as fork. Under MPI we can't share any memory between nodes because they are likely different hosts.

> (after we release 2.0) we should rework `wordlist.c` and just seek to offset on restore, like core always did.

Agreed.
